### PR TITLE
transform helper

### PR DESCRIPTION
**Pull request by jryurkanin**
The helper function parses the run properties and results data so that the user doesn't need to include any parsing script in their transform script. The helper function only accepts a user-defined function that accepts an array of results data as its input and returns an array of results data matching the assay design it is used with. Also, the helper function requires that users explicitly define the run properties file path in their transform script, otherwise the labkey transform script parser does not recognize ${runInfo}.

#### Rationale

Users currently must write the parsing script for run properties and results data every time they make a transform script for assays. Because the parsing script will almost always be the same, this helper function can be used to save users time instead.

#### Changes
* add helper function for parsing run properties and results-level data in an assay transform script.

### CODE REVIEW

**Comment by jryurkanin** [comment]
i'll do another PR at some point to broaden this so it handles well plates better. right now its dependent on labkey pre-parsing results data in to a tsv.

**Review by cnathe — COMMENTED** [review]
These run property type helpers are great! Thank you for pushing this forward. I'll defer to Alan on the actual code review.

File: samples/assay_transform_function_example.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# Import the transform helper from the labkey python api library
+from labkey.utils import transform_helper
+
+# the run properties filepath must be defined in this way so that it can be fed into the transform helper
+filepath = '${runInfo}'
+
+# define the function that you are using to transform your results data array
+def transform(grid):
```
Comment (labkey-alan) [inline samples/assay_transform_function_example.py]: I am trying to move us away from this samples folder, can you move this code example to a markdown file in the `/docs` folder?

File: labkey/utils.py [hunk 2]
```diff
@@ -31,3 +31,44 @@ def default(self, o):
 def json_dumps(*args, **kwargs):
     kwargs.setdefault("cls", DateTimeEncoder)
     return json.dumps(*args, **kwargs)
+
+
+def transform_helper(userTransformFunc, filePathRunProperties):
+    #filePathRunProperties must be explicitly defined as ${runInfo} within the user's transform script
+    #parse run properties to for results data in and out filepaths
+    filePathIn = ''
```
Comment (labkey-alan) [inline labkey/utils.py]: In Python the generally accepted naming convention is to use `snake_case` for variable names, not `camelCase`, and `PascalCase` for classes. Can you convert these variables to be `snake_case`?

**Review by labkey-alan — CHANGES_REQUESTED** [review]
This is mostly ready, but we should move the example code to the docs folder, and make the code style more Pythonic (and aligned with what we do in this repo).

**Comment by jryurkanin** [comment]
@labkey-alan okay, should be good now. Let me know if I can make it match our style better... I know I need to be more mindful to keep python scripts pythonic.

File: labkey/utils.py [hunk 3]
```diff
@@ -31,3 +31,45 @@ def default(self, o):
...
+    file_path_in = ''
+    file_path_out = ''
+    file_run_properties = open(file_path_run_properties) 
+    for l in file_run_properties:
+        row = l.strip().split('\t')
+        if row[0] == 'runDataFile':
+            file_path_out = row[3]
+        if row[0] == 'runDataUploadedFile':
+            file_path_in = row[1]
+    file_run_properties.close()
```
Comment (labkey-alan) [inline labkey/utils.py]:
```suggestion
    with open(file_path_run_properties) as file_run_properties:
        for l in file_run_properties:
            row = l.strip().split('\t')
            if row[0] == 'runDataFile':
                file_path_out = row[3]
            if row[0] == 'runDataUploadedFile':
                file_path_in = row[1]
```

Comment (labkey-alan) [reply labkey/utils.py]: When using open it's best to use it with a `with` statement, that way if any errors happen before close the file handle automatically gets closed.

File: labkey/utils.py [hunk 4]
```diff
@@ -31,3 +31,45 @@ def default(self, o):
...
+    
+    #parse results data into array, confirming supported file type is used
+    file_in = open(file_path_in)
```
Comment (labkey-alan) [inline labkey/utils.py]: As above, we should convert this to a with statement

File: labkey/utils.py [hunk 5]
```diff
@@ -31,3 +31,45 @@ def default(self, o):
...
+            raise ValueError('Unsupported file type or delimiter used. Header used: \n' + str(l))
+        data_grid.append(row)    
+    file_in.close()
+    
+    #run user transform on parsed results data array
+    transformed_grid = user_transform_func(data_grid)
+    
+    #write transformed results data array to LabKey assay results data grid
+    #transformed array must be a python list object, not a numpy array or pandas dataframe
+    file_out = open(file_path_out, mode='w')
```
Comment (labkey-alan) [inline labkey/utils.py]: As above we should convert this to a with statement

**Review by labkey-alan — CHANGES_REQUESTED** [review]
Sorry for the delay in reviewing! Let's clean up these open statements then we can merge and I'll do a release with these changes.

File: labkey/utils.py (hunk 3 above)
Comment (jryurkanin) [reply labkey/utils.py]: got it, makes sense. in retrospect, I've only rarely seen files opened without a with statement.

**Review by labkey-alan — APPROVED** [review]
Looks good!

**Pull request merged**